docs/GDPs/gdp-0003-duck-storage.rst

```diff
@@ -0,0 +1,822 @@
+:code:`np.negative(storage)`) alternatively, some ufuncs can be used to perform reductions.
+In this case, one can explicitly call the :code:`reduce` method (e.g. :code:`np.add.reduce(axis=1)`
+to accumulate the values along a given axis.
+
+We propose to support the methods :code:`__call__` and :code:`reduce` of the NumPy ufunc
+mechanism.
+
+If the :code:`reduce` method of `ufuncs` is used, this results in a Storage with the axis along
+which the reduction was performed removed from the Storage. (For example taking the sum over the K
+axis of an IJK storage will result in an IJ storage). In addition to vanilla numpy behavior, the
+:code:`axis` keyword of the :code:`reduce` method accepts the axis along which the reduction is
+performed as a string.
+
+In the following subsections we describe the proposed behaviour for GT4Py storages when used in
+conjuction with NumPy ufuncs.
+
+.. _broadcasting:
+
+Broadcasting
+============
+
+With the term "broadcasting", NumPy describes the ways that different shapes are combined in
+assignments and mathematical operations. GT4Py storages should override the default NumPy behavior
+so that fields are broadcast along the same spatial dimension. For example, adding an :code:`IJ`
+field :code:`A` of shape :code:`(2, 3)` with a :code:`K` field :code:`B` of shape :code:`(4,)` will
+result in an :code:`IJK` field :code:`C` of shape :code:`(2, 3, 4)`, with `C[i,j,k] = A[i,j]+B[k]`.
+
+Similarly, fields of lower dimension are assigned to such of higher dimension by broadcasting along
+the missing axes. To keep compatibility with NumPy, dimensions of size 1 are treated like missing
+axes when broadcasting.
+
+Similarly, fields of lower dimension are assigned to such of higher dimension by broadcasting along
+the missing dimensions. To keep compatibility with NumPy, dimensions of size 1 are treated like
+missing axes when broadcasting.
+
+
+.. _output_storage_parameters:
+
+Output Storage Parameters
+=========================
+
+If no output buffer is provided, the constructor parameters of the output storage have to be
+inferred using the available information from the inputs.
+
+:code:`aligned_index`
+    It is chosen to be as the largest value per dimension across all inputs which are a GT4Py
+    Storage.
+
+:code:`alignment`
+   The resulting alignment is chosen as the least common multiple of the alignments of all inputs
+   which are a GT4Py storage.
+
+:code:`axes`
+    If the :code:`axes` parameters of all operands agree, the output will have the same
+    :code:`axes`. Otherwise, if one input contains all axes of the other input, that input
+    determines the axes of the output. If none of these conditions are met, axes are chosen as the
+    union of all input storages. Their order will then be a sub-sequence of "IJK".
+
+:code:`dtype`
+   The resulting dtype is determined by NumPy behavior.
+
+:code:`halo`
+    It is chosen s.t. the resulting domain is the intersection of all individual domains.
```

Comment:
If you e.g. add two fields, this will need the allocation of a new storage where the result will be stored. Now what should the halo be of that resulting storage, if the halo of the two input storages are not the same? How we resolve this conflict is by saying that the grid points that were in the inner domain for _all_ the input storages will be exactly the ones in the inner domain of the resulting one.